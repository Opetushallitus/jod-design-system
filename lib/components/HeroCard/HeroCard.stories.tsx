import type { Meta, StoryObj } from '@storybook/react';

import { HeroCard } from './HeroCard';

const meta = {
  title: 'Cards/HeroCard',
  component: HeroCard,
  tags: ['autodocs'],
} satisfies Meta<typeof HeroCard>;

export default meta;

type Story = StoryObj<typeof meta>;

const secondaryCardArgs = {
  title: 'Tunnista osaamistasi ja tutki mahdollisuuksia',
  backgroundColor: '#00A8B3F2',
  href: '/',
  size: 'sm' as 'sm' | 'lg' | undefined,
  textColor: '#000',
};

export const Primary: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2225-24513',
    },
    docs: {
      description: {
        story: 'This is a primary hero card component.',
      },
    },
    backgrounds: {
      default: 'light',
    },
  },
  args: {
    title: 'Astu rohkeasti muutokseen! Tutki mahdollisuuksiasi, suunnittele polkusi ja luo tulevaisuutesi.',
    content:
      'Elämä on jatkuvaa muutosta ja uudistumista. Olipa kyse uran vaihdosta, uuden osaamisen oppimisesta tai henkilökohtaisen elämäntilanteen parantamisesta, me tarjoamme sinulle tiedon ja inspiraation joita tarvitset.',
    backgroundColor: '#006DB3F2',
    actionContent: 'Kokeile palvelua',
    arrowColor: '#006DB3',
    href: '/',
    onClick: () => console.log('Clicked'),
  },
  decorators: [
    (Story) => (
      <div className="ds-mb-[30px] ds-max-w-2xl">
        <Story />
      </div>
    ),
  ],
};

export const PrimaryWithoutArrowsOrActions: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2225-24513',
    },
    docs: {
      description: {
        story: 'This is a primary hero card component without arrows or actions.',
      },
    },
    backgrounds: {
      default: 'light',
    },
  },
  args: {
    title: 'Amet sit dolor ipsum lorem',
    content:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam nec dui ac nunc tincidunt tincidunt. Nullam nec dui ac nunc tincidunt tincidunt.',
    backgroundColor: '#006DB3F2',
    arrowVisible: false,
    href: '/',
  },
  decorators: [
    (Story) => (
      <div className="ds-mb-[30px] ds-max-w-2xl">
        <Story />
      </div>
    ),
  ],
};

export const Secondary: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=541%3A4604',
    },
    docs: {
      description: {
        story: 'This is a secondary hero card component.',
      },
    },
  },
  args: secondaryCardArgs,
  decorators: [
    (Story) => (
      <div className="ds-max-w-xs">
        <Story />
      </div>
    ),
  ],
};

export const MultipleSecondary: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=541%3A4574',
    },
    docs: {
      description: {
        story: 'Multiple hero card components.',
      },
    },
  },
  args: secondaryCardArgs,
  decorators: [
    (Story) => (
      <div className="ds-flex ds-flex-row ds-gap-[32px] lg:ds-container">
        <Story />
        <Story
          args={{
            ...secondaryCardArgs,
            title: 'Tutustu miten käytät palvelua ja luot oman tulevaisuutesi',
            backgroundColor: '#EE7C45F2',
          }}
        />
        <Story
          args={{
            ...secondaryCardArgs,
            title: 'Luo oma profiili ja suunnittele polkusi',
            backgroundColor: '#CD4EB3F2',
          }}
        />
      </div>
    ),
  ],
};

export const Hero: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=541%3A4574',
    },
    docs: {
      description: {
        story: 'This is a hero component with multiple hero card components.',
      },
    },
    backgrounds: {
      default: 'light',
    },
    layout: 'fullscreen',
  },
  args: {
    title: 'Astu rohkeasti muutokseen! Tutki mahdollisuuksiasi, suunnittele polkusi ja luo tulevaisuutesi.',
    content:
      'Elämä on jatkuvaa muutosta ja uudistumista. Olipa kyse uran vaihdosta, uuden osaamisen oppimisesta tai henkilökohtaisen elämäntilanteen parantamisesta, me tarjoamme sinulle tiedon ja inspiraation joita tarvitset.',
    backgroundColor: '#006DB3F2',
    actionContent: 'Kokeile palvelua',
    href: '/',
    arrowColor: '#006DB3',
  },
  decorators: [
    (Story) => (
      <div className="ds-mx-auto ds-flex ds-flex-col ds-gap-[104px] ds-bg-[url('https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=2070')] ds-bg-[length:2000px_cover] ds-bg-[top_-6rem_left_-10rem] ds-p-8 lg:ds-container">
        <div className="ds-mb-[30px] ds-max-w-2xl">
          <Story />
        </div>
        <div className="ds-flex ds-flex-row ds-gap-[32px] lg:ds-container">
          <Story args={secondaryCardArgs} />
          <Story
            args={{
              ...secondaryCardArgs,
              title: 'Tutustu miten käytät palvelua ja luot oman tulevaisuutesi',
              backgroundColor: '#EE7C45F2',
            }}
          />
          <Story
            args={{
              ...secondaryCardArgs,
              title: 'Luo oma profiili ja suunnittele polkusi',
              backgroundColor: '#CD4EB3F2',
            }}
          />
        </div>
      </div>
    ),
  ],
};
