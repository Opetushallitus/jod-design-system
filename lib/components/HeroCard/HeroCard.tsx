export interface HeroCardProps {
  /** Title text shown on the card */
  title: string;
  /** Content text shown on the card */
  content?: string;
  /** Action content shown on the card */
  actionContent?: string;
  /** Background color of the card */
  backgroundColor?: string;
  /** Link to page */
  href?: string;
  /** Text color */
  textColor?: string;
  /** Size of the card. Lg is default. */
  size?: 'lg' | 'sm';
  /** Arrow visibility. By default it's true if href or actionContent is present. */
  arrowVisible?: boolean;
  /** Arrow color. Defaults to text color */
  arrowColor?: string;
  /** Override the default link click event handling */
  onClick?: () => void;
}

const Arrow = ({ color }: { color: string }) => (
  <div className="size-32 ml-auto select-none font-bold" aria-hidden>
    <svg xmlns="http://www.w3.org/2000/svg" width="48" height="48" viewBox="0 0 48 48" fill="none">
      <path d="M30 38L27.15 35.2L36.35 26H4V22H36.35L27.2 12.8L30 10L44 24L30 38Z" fill={color} />
    </svg>
  </div>
);

/** Cards group information into flexible containers that allow users to browse a collection of related items and actions. */
export const HeroCard = ({
  title,
  content,
  actionContent,
  backgroundColor = '#444BACF2',
  href,
  textColor = '#FFF',
  arrowColor = textColor,
  arrowVisible = !!href || !!actionContent,
  size = 'lg',
  onClick,
}: HeroCardProps) => {
  const handleLinkClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
    if (onClick) {
      e?.preventDefault?.();
      onClick();
    }
  };

  if (size === 'lg') {
    return (
      <div
        className="font-poppins relative flex flex-col gap-6 rounded-lg px-7 pb-11 pt-7 tracking-widest"
        style={{ backgroundColor, color: textColor }}
      >
        <h1 className="overflow text-pretty text-[36px] font-semibold leading-[50px]">{title}</h1>
        {content && <p className="tracking-normal leading-[27px] font-medium text-[18px]">{content}</p>}
        {actionContent && (
          <a
            href={href}
            className="absolute bottom-0 right-7 translate-y-2/4 rounded-[40px] outline-none transition-transform hover:scale-105 focus:scale-105"
            onClick={handleLinkClick}
          >
            <div
              className="flex select-none items-center gap-4 rounded-[40px] bg-white px-8 py-[20px] text-[28px] font-bold leading-[40px]"
              style={{ color: backgroundColor }}
            >
              {actionContent}
              {arrowVisible && <Arrow color={arrowColor} />}
            </div>
          </a>
        )}
        {!actionContent && arrowVisible && <Arrow color={arrowColor} />}
      </div>
    );
  }

  if (size === 'sm') {
    return (
      <a
        href={href}
        className="flex rounded-lg outline-none transition-transform hover:scale-105 focus:scale-105"
        onClick={handleLinkClick}
      >
        <div
          className="font-poppins relative flex flex-col grow gap-6 rounded-lg p-7 pb-6 tracking-widest justify-between"
          style={{ backgroundColor, color: textColor }}
        >
          <h2 className="overflow text-pretty text-[24px] font-semibold leading-[34px]">{title}</h2>
          {arrowVisible && <Arrow color={arrowColor} />}
        </div>
      </a>
    );
  }
};
