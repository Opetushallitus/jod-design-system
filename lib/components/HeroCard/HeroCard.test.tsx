import '@testing-library/jest-dom';
import { createEvent, fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { HeroCard, HeroCardProps } from './HeroCard';

describe('HeroCard', () => {
  const title = 'Test Title';
  const content = 'Test Content';
  const actionContent = 'Test Action Content';
  const backgroundColor = '#444BACF2';

  it('renders HeroCard with actionContent', () => {
    render(
      <HeroCard
        to="/"
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        title={title}
        content={content}
        actionContent={actionContent}
        backgroundColor={backgroundColor}
      />,
    );

    // Assert title, content, and actionContent are rendered
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(content)).toBeInTheDocument();
    expect(screen.getByText(actionContent)).toBeInTheDocument();
  });

  it('renders HeroCard with a link when actionContent is provided', () => {
    render(
      <HeroCard
        to="/"
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        title={title}
        content={content}
        actionContent={actionContent}
        backgroundColor={backgroundColor}
      />,
    );

    // Assert HeroCard is wrapped in a link
    const linkElement = screen.getByRole('link');
    expect(linkElement).toBeInTheDocument();
  });

  it('renders HeroCard without actionContent', () => {
    render(
      <HeroCard
        to="/"
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        title={title}
        content={content}
        backgroundColor={backgroundColor}
      />,
    );

    // Assert title and content are rendered
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(content)).toBeInTheDocument();

    // Assert actionContent is not rendered
    expect(screen.queryByText(actionContent)).toBeNull();
  });

  it('renders the arrow by default if to and linkComponent and actionContent are present', () => {
    const defaultProps: HeroCardProps = {
      title,
      content,
      backgroundColor,
    };
    const { rerender } = render(<HeroCard {...defaultProps} />);

    // Assert arrow is not rendered
    expect(document.querySelector('svg')).toBeNull();

    // Assert arrow is rendered when href is provided
    rerender(
      <HeroCard
        {...defaultProps}
        to={'/'}
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
      />,
    );
    expect(document.querySelector('svg')).not.toBeNull();

    // Assert arrow is rendered when actionContent is provided
    rerender(
      <HeroCard
        {...defaultProps}
        to={'/'}
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        actionContent={actionContent}
      />,
    );
    expect(document.querySelector('svg')).not.toBeNull();
  });

  it('renders HeroCard with a link when size is sm', () => {
    render(
      <HeroCard
        to="/"
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        title={title}
        content={content}
        backgroundColor={backgroundColor}
        size="sm"
      />,
    );

    // Assert HeroCard is wrapped in a link
    const linkElement = screen.getByRole('link');
    expect(linkElement).toBeInTheDocument();
  });

  it('handles link clicks normally when onClick is not provided', () => {
    render(
      <HeroCard
        to="/"
        linkComponent={({ to, children }) => <a href={to as string}>{children}</a>}
        title={title}
        content={content}
        backgroundColor={backgroundColor}
        size="sm"
      />,
    );

    // Assert HeroCard is wrapped in a link
    const linkElement = screen.getByRole('link');
    expect(linkElement).toBeInTheDocument();
    const clickEvent = createEvent.click(linkElement);
    fireEvent(linkElement, clickEvent);
    expect(clickEvent.defaultPrevented).toBe(false);
  });
});
