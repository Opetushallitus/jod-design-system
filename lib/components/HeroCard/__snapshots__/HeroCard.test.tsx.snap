// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeroCard > renders HeroCard with a button 1`] = `
<div
  class="ds:flex ds:flex-col ds:gap-4 ds:rounded-lg ds:p-6 ds:justify-between ds:text-white ds:hyphens-auto"
  style="background-color: rgba(68, 75, 172, 0.95);"
>
  <h2
    class="ds:text-pretty ds:mr-9 ds:text-hero-mobile ds:sm:text-hero"
  >
    Test Title
  </h2>
  <p
    class="ds:text-pretty ds:text-body-lg-mobile ds:sm:text-body-lg"
  >
    Test Content
  </p>
  <div>
    <a
      href="/"
    >
      <span
        class="ds:cursor-pointer ds:inline-flex ds:items-center ds:gap-4 ds:select-none ds:group ds:font-poppins ds:text-button-md ds:px-6 ds:rounded-[30px] ds:pr-4 ds:text-primary-gray ds:bg-white ds:hover:text-accent ds:focus-visible:text-accent ds:active:bg-accent ds:focus-visible:outline-accent ds:focus-visible:outline ds:focus-visible:outline-[3px] ds:focus-visible:outline-offset-[1.5px] ds:active:text-white ds:active:outline-0 ds:mt-4 ds:w-fit ds:group-focus:underline ds:group-focus:text-accent"
      >
        <span
          class="ds:group-hover:underline ds:group-active:no-underline ds:group-focus-visible:no-underline ds:py-4"
        >
          Test Button Label
        </span>
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`HeroCard > renders HeroCard without a button 1`] = `
<div
  class="ds:flex ds:flex-col ds:gap-4 ds:rounded-lg ds:p-6 ds:justify-between ds:text-white ds:hyphens-auto"
  style="background-color: rgba(68, 75, 172, 0.95);"
>
  <h2
    class="ds:text-pretty ds:mr-9 ds:text-hero-mobile ds:sm:text-hero"
  >
    Test Title
  </h2>
  <p
    class="ds:text-pretty ds:text-body-lg-mobile ds:sm:text-body-lg"
  >
    Test Content
  </p>
</div>
`;
